internal/mining/cpuminer/cpuminer.go

```diff
@@ -490,205 +543,227 @@ out:
-// NumWorkers returns the number of workers which are running to solve blocks.
+// NumWorkers returns the number of workers which are running to solve blocks
+// in the normal mining mode.
 //
 // This function is safe for concurrent access.
 func (m *CPUMiner) NumWorkers() int32 {
 	return int32(atomic.LoadUint32(&m.numWorkers))
 }
 
-// GenerateNBlocks generates the requested number of blocks. It is self
-// contained in that it creates block templates and attempts to solve them while
-// detecting when it is performing stale work and reacting accordingly by
-// generating a new block template.  When a block is solved, it is submitted.
-// The function returns a list of the hashes of generated blocks.
+// GenerateNBlocks generates the requested number of blocks in the discrete
+// mining mode and returns a list of the hashes of generated blocks that were
+// added to the main chain.
+//
+// It makes use of a subscription to the background block template generator to
+// obtain the templates and attempts to solve them while automatically switching
+// to new templates as they become available as needed.  As a result, it
+// supports many of the nice features of the template subscriptions such as
+// giving all votes a chance to arrive.
+//
+// Note that, as the above implies, this will only consider blocks successfully
+// added to the main chain in the overall count, so, upon returning, the list of
+// hashes will only contain the hashes of those blocks.  This distinction is
+// important because it is sometimes possible for a block to be rejected or be
+// added to a side chain if it happens to be solved around the same time another
+// one shows up.
 func (m *CPUMiner) GenerateNBlocks(ctx context.Context, n uint32) ([]*chainhash.Hash, error) {
+	// Nothing to do.
+	if n == 0 {
+		return nil, nil
+	}
+
 	// Respond with an error if server is already mining.
 	m.Lock()
-	if m.started || m.discreteMining {
+	if m.normalMining || m.discreteMining {
 		m.Unlock()
 		return nil, errors.New("server is already CPU mining. Please call " +
 			"`setgenerate 0` before calling discrete `generate` commands")
 	}
 
-	m.started = true
 	m.discreteMining = true
 	m.Unlock()
 
 	log.Tracef("Generating %d blocks", n)
 
-	i := uint32(0)
-	blockHashes := make([]*chainhash.Hash, n)
-
 	// Start a ticker which is used to signal checks for stale work.
 	ticker := time.NewTicker(time.Second * hashUpdateSecs)
 	defer ticker.Stop()
 
+	templateSub := m.g.Subscribe()
+	defer templateSub.Stop()
+
+	blockHashes := make([]*chainhash.Hash, 0, n)
+	var stats speedStats
+out:
 	for {
-		// Read updateNumWorkers in case someone tries a `setgenerate` while
-		// we're generating. We can ignore it as the `generate` RPC call only
-		// uses 1 worker.
+		// Wait for a new template update notification or early shutdown.
+		var templateNtfn *mining.TemplateNtfn
 		select {
-		case <-m.updateNumWorkers:
-		default:
+		case <-ctx.Done():
+			break out
+		case <-m.quit:
+			break out
```

Comment:
I wanted to be explicit here because this context is not the same one as the one used to start the CPU miner.  The CPU miner is started via the server, but this method is called via the RPC server with the context of the connection to the RPC server itself. 

The quit channel is closed when the server context is cancelled (aka typically when the server is shutdown) and this context is cancelled when the specific RPC connection that invoked the method goes away.  That way, if the caller tells it to generate a bunch of blocks and then closes the connection (e.g. hits ctrl+c after doing `dcrctl generate 1000`) this loop will exit, and it will also exit if the CPU miner is shutdown otherwise for any reason.

In reality, the server shutting down will also shutdown the RPC server which in turn will disconnect all of the clients and that would end up cancelling the context here, so we could technically get away with only listening for the context cancellation here, but that would mean this code would heavily rely on the ordering of subsystems and also have to make a bunch of assumptions about the way the other subsystems work.  Such assumptions would make this code brittle and introduce a bunch of hidden state that nobody could reasonably know about when working on the other subsystems which I want to avoid.

Reply:
I wanted to be explicit here because this context is not the same one as the one used to start the CPU miner.  The CPU miner is started via the server, but this method is called via the RPC server with the context of the connection to the RPC server itself. 

The quit channel is closed when the server context is cancelled (aka typically when the server is shutdown) and this context is cancelled when the specific RPC connection that invoked the method goes away.  That way, if the caller tells it to generate a bunch of blocks and then closes the connection (e.g. hits ctrl+c after doing `dcrctl generate 1000`) this loop will exit, and it will also exit if the CPU miner is shutdown otherwise for any reason.

In reality, the server shutting down will also shutdown the RPC server which in turn will disconnect all of the clients and that would end up cancelling the context here, so we could technically get away with only listening for the context cancellation here, but that would mean this code would heavily rely on the ordering of subsystems and also have to make a bunch of assumptions about the way the other subsystems work.  Such assumptions would make this code brittle and introduce a bunch of hidden state that nobody could reasonably know about when working on the other subsystems which I want to avoid.

Reply:
I see now. This is a completely different context than the one running through `Run(ctx ...)` and comes from the RPC call. It is totally needed then.